clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/model/EcsServerGroup.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright 2017 Lookout, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.ecs.model;
+
+import com.netflix.spinnaker.clouddriver.model.Instance;
+import com.netflix.spinnaker.clouddriver.model.ServerGroup;
+
+import java.util.Map;
+import java.util.Set;
+
+public class EcsServerGroup implements ServerGroup {
+
+  String name;
+  String type;
+  String cloudProvider;
+  String region;
+  Boolean isDisabled;
+  Long createdTime;
+  Set<String> zones;
+  Set<Instance> instances;
+  Set<String> loadBalancers;
+  Set<String> securityGroups;
+  Map<String, Object> launchConfig;
+  InstanceCounts instanceCounts;
+  Capacity capacity;
+  ImagesSummary imagesSummary;
+  ImageSummary imageSummary;
+
+  public EcsServerGroup() {
+
+  }
+
+  @Override
+  public String getName() {
+    return name;
+  }
+
+  @Override
+  public String getType() {
+    return type;
+  }
+
+  @Override
+  public String getCloudProvider() {
+    return cloudProvider;
+  }
+
+  @Override
+  public String getRegion() {
+    return region;
+  }
+
+  public Boolean getDisabled() {
+    return isDisabled;
+  }
+
+  @Override
+  public Long getCreatedTime() {
+    return createdTime;
+  }
+
+  @Override
+  public Set<String> getZones() {
+    return zones;
+  }
+
+  @Override
+  public Set<Instance> getInstances() {
+    return instances;
+  }
+
+  @Override
+  public Set<String> getLoadBalancers() {
+    return loadBalancers;
+  }
+
+  @Override
+  public Set<String> getSecurityGroups() {
+    return securityGroups;
+  }
+
+  @Override
+  public Map<String, Object> getLaunchConfig() {
+    return launchConfig;
+  }
+
+  @Override
+  public InstanceCounts getInstanceCounts() {
+    return instanceCounts;
+  }
+
+  @Override
+  public Capacity getCapacity() {
+    return capacity;
+  }
+
+  @Override
+  public ImagesSummary getImagesSummary() {
+    return imagesSummary;
+  }
+
+  @Override
+  public ImageSummary getImageSummary() {
+    return imageSummary;
+  }
+
+  @Override
+  public Boolean isDisabled() {
+    return null;
+  }
+
+  @Override
+  public Map<String, Object> getTags() {
+    return null;
+  }
+
+  public EcsServerGroup withName(String name) {
```

Comment:
what is the purpose of the withXYZ(parameter) construct, vs just using setters, or in this case initializing all of these values in the constructor?

Reply:
withXYZ(parameter) returns the object being modified, which allows you to chain up multiple calls, eg. new EcsServerGroup().withName("my-cool-name").withRegion("us-west-2");  This is less verbose, and easier to read

A constructor with multiple parameters is a no-no in java.  See:
```n his book Code Complete, Steve McConnell argues that no procedure should ever have more than a maximum of 6, maybe 7 arguments. Most of these statements are not just his opinion but backed by studies, e.g. of error rates related to code structure.

Clean Code by Robert Martin goes even further: He recommends 1 or 2 arguments, while 3 is already considered a "code smell". Personally, I think Clean Code is a bit extreme in places, but on the whole it makes some good arguments.

"A whole bunch of parameters" (however many that may be) bespeaks a "kitchen sink" design with lots of afterthoughts and little structure. It also makes maintenance more difficult and error prone; at the very least, it makes code hard to read.

All this makes for good reasons to think about cutting down on the number of parameters. Other answers have offered some practical suggestions.```
https://stackoverflow.com/questions/1632058/java-constructor-with-large-arguments-or-java-bean-getter-setter-approach

Reply:
In some cases, having a proper builder pattern makes sense, and sometimes it is an overkill (e.g. this one where we just have a model class)